### ISSUE TITLE
space-before-function-paren doesn't work properly

### ISSUE THREAD

**Issue by FezVrasta** [eslint/eslint#6328]
**What version of ESLint are you using?**
eslint@2.10.1

**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint@6.0.4

**Please show your full configuration:**
```js
  "eslintConfig": {
    "parser": "babel-eslint",
    "extends": [
      "eslint:recommended",
      "plugin:react/recommended",
      "airbnb"
    ],
    "env": {
      "browser": true,
      "node": true,
      "mocha": true,
      "es6": true
    },
    "plugins": [
      "babel",
      "react",
      "jsx-a11y"
    ],
    "parserOptions": {
      "ecmaVersion": 6,
      "sourceType": "module",
      "ecmaFeatures": {
        "jsx": true
      }
    },
    "rules": {
      "import/no-unresolved": "error",
      "comma-dangle": [
        "error",
        "always-multiline"
      ],
      "indent": [
        "error",
        2,
        {
          "SwitchCase": 1
        }
      ],
      "spaced-comment": [
        "error",
        "always",
        {
          "markers": [
            "#"
          ]
        }
      ],
      "new-cap": "off",
      "babel/new-cap": "warn",
      "quotes": [
        "error",
        "single",
        {
          "avoidEscape": true,
          "allowTemplateLiterals": true
        }
      ],
      "jsx-quotes": [
        "error",
        "prefer-single"
      ],
      "no-console": "warn",
      "no-debugger": "warn",
      "no-restricted-syntax": "off",
      "max-len": "off",
      "prefer-template": "error",
      "no-use-before-define": "off",
      "newline-per-chained-call": "off",
      "arrow-body-style": [
        "error",
        "as-needed"
      ],
      "react/jsx-indent": [
        "error",
        2
      ],
      "react/jsx-indent-props": [
        "error",
        2
      ],
      "react/jsx-first-prop-new-line": [
        "error",
        "multiline"
      ],
      "react/prefer-stateless-function": "off",
      "jsx-a11y/href-no-hash": "error",
      "jsx-a11y/label-has-for": "error",
      "jsx-a11y/mouse-events-have-key-events": "error",
      "jsx-a11y/role-has-required-aria-props": "error",
      "jsx-a11y/role-supports-aria-props": "error",
      "jsx-a11y/aria-props": "error"
    },
    "settings": {
      "import/resolver": {
        "webpack": {
          "config": "./internals/webpack/webpack.test.babel.js"
        }
      }
    }
  },
```

**What did you do? Please include the actual source code causing the issue.**
```js
/* eslint prefer-arrow-callback: "off" */
/* eslint func-names: "off" */
export class Foo {
  fun () { // space-before-function-paren error (unexpected space)
    return function() {  // space-before-function-paren error (missing space)
      return true;
    };
  }
}
```

**What did you expect to happen?**
The `return function() {` line shouldn't throw an error

**What actually happened? Please include the actual, raw output from ESLint.**
The second line throws an error (correctly)
The third line, throws an error (but it's in contrast with the error of the line above)

![image](https://cloud.githubusercontent.com/assets/5382443/15834539/411fdd54-2c2c-11e6-8af5-645dd78f4eb4.png)

**Comment by btmills** [comment]
AirBnB's config, which you're extending without overriding, has `space-before-function-paren` [configured](https://github.com/airbnb/javascript/blob/f41b4df02458638525779c4ce3e496a9b588ebe5/packages/eslint-config-airbnb-base/rules/style.js#L232) as `[2, { anonymous: 'always', named: 'never' }]`. Since the `return function() {` line is an anonymous function, the rule is correctly expecting a space there. It looks like everything is behaving as expected. @FezVrasta am I reading everything correctly?

**Comment by FezVrasta** [comment]
Thanks, I didn't noticed that rule :(
